airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3Client.kt

```diff
@@ -47,30 +68,139 @@ class S3Client(
         }
     }
 
-    override suspend fun put(key: String, bytes: ByteArray) {
+    override suspend fun move(remoteObject: S3Object, toKey: String): S3Object {
+        val request = CopyObjectRequest {
+            bucket = bucketConfig.s3BucketName
+            this.key = toKey
+            copySource = remoteObject.keyWithBucketName
+        }
+        client.copyObject(request)
+        delete(remoteObject)
+        return S3Object(toKey, bucketConfig)
+    }
+
+    override suspend fun <R> get(key: String, block: (InputStream) -> R): R {
+        val request = GetObjectRequest {
+            bucket = bucketConfig.s3BucketName
+            this.key = key
+        }
+        return client.getObject(request) {
+            it.body?.toInputStream()?.use(block)
+                ?: throw IllegalStateException(
+                    "S3 object body is null (this indicates a failure, not an empty object"
+                )
+        }
+    }
+
+    override suspend fun put(key: String, bytes: ByteArray): S3Object {
         val request = PutObjectRequest {
             bucket = bucketConfig.s3BucketName
             this.key = key
             body = ByteStream.fromBytes(bytes)
         }
         client.putObject(request)
+        return S3Object(key, bucketConfig)
     }
 
-    override suspend fun delete(key: String) {
-        val request =
-            aws.sdk.kotlin.services.s3.model.DeleteObjectRequest {
-                bucket = bucketConfig.s3BucketName
-                this.key = key
-            }
+    override suspend fun delete(remoteObject: S3Object) {
+        val request = DeleteObjectRequest {
+            bucket = remoteObject.bucketConfig.s3BucketName
+            this.key = remoteObject.key
+        }
         client.deleteObject(request)
     }
+
+    override suspend fun streamingUpload(
+        key: String,
+        collector: suspend S3MultipartUpload.Writer.() -> Unit
+    ): S3Object {
+        val request = CreateMultipartUploadRequest {
+            this.bucket = bucketConfig.s3BucketName
+            this.key = key
+        }
+        val response = client.createMultipartUpload(request)
+        S3MultipartUpload(response).upload(collector)
```

Comment:
Hm... I'll look at it. I was trying to keep it constrained to what was necessary to reduce the number of variables

Reply:
So, doing something like

```
fun streamingUpload(key: String): OutputStream
```
where
```
class S3MultipartUpload( ... ): OutputStream {
  override fun write(int n) {
    // write to underlying buffer
    // check size
    // maybe uploadPart()
```
almost works, but doesn't, because `write` needs to be a suspend fun (because `uploadPart` needs to be)

Also if you just take and return the output stream, letting the client close it, there's no way to return an S3Object when you're done (not a dealbreaker obv, but it's less than ideal).

But even this still has the issue with `write`.
```
fun streamingUpload(key: String, block: (OutputStream) -> Unit): S3Object
```

I will take 2 out of 3 and change it to this:

```
override suspend fun streamingUpload(
        key: String,
        block: suspend (S3MultipartUpload.Writer) -> Unit
    ): S3Object {
        val request = CreateMultipartUploadRequest {
            this.bucket = bucketConfig.s3BucketName
            this.key = key
        }
        val response = client.createMultipartUpload(request)
        log.info {
            "Starting multipart upload to ${response.bucket}/${response.key} (${response.uploadId}"
        }
        val upload = S3MultipartUpload(response)
        block(upload.Writer())
        upload.complete()
        return S3Object(key, bucketConfig)
    }
```

However, I'd prefer to make this change in the compression PR, because it adds the extra complexity of a compression processor, so I want to make I don't break stuff.